services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_aggregate_query.rb

```diff
@@ -0,0 +1,150 @@
+# frozen_string_literal: true
+
+module AdminDiagnosticReports
+  class DiagnosticAggregateQuery < Snapshots::PeriodQuery
+    class InvalidAggregationError < StandardError; end
+
+    attr_reader :additional_aggregation
+
+    AGGREGATION_OPTIONS = [
+      'grade',
+      'teacher',
+      'classroom'
+    ]
+    DIAGNOSTIC_ORDER_BY_ID = [
+      1663, # Starter Baseline Diagnostic (Pre)
+      1668, # Intermediate Baseline Diagnostic (Pre)
+      1678, # Advanced Baseline Diagnostic (Pre)
+      1161, # ELL Starter Baseline Diagnostic (Pre)
+      1568, # ELL Intermediate Baseline Diagnostic (Pre)
+      1590  # ELL Advanced Baseline Diagnostic (Pre)
+    ]
+
+    def initialize(aggregation:, **options)
+      raise InvalidAggregationError, "#{aggregation} is not a valid aggregation value." unless AGGREGATION_OPTIONS.include?(aggregation)
+
+      @additional_aggregation = aggregation
+
+      super(**options)
+    end
+
+    def run
+      post_process(run_query)
+    end
+
+    def select_clause
+      <<-SQL
+        SELECT
+          activities.id AS diagnostic_id,
+          activities.name AS diagnostic_name,
+          #{aggregation_clause} AS name,
+          #{specific_select_clause}
+      SQL
+    end
+
+    def from_and_join_clauses
+      super + <<-SQL
+        JOIN lms.classroom_units
+          ON classrooms.id = classroom_units.classroom_id
+        JOIN lms.users
+          ON schools_users.user_id = users.id
+      SQL
+    end
+
+    def specific_select_clause
+      raise NotImplementedError
+    end
+
+    def where_clause
+      super + <<-SQL
+          #{relevant_diagnostic_where_clause}
+      SQL
+    end
+
+    def relevant_diagnostic_where_clause
+      "AND activities.id IN (#{DIAGNOSTIC_ORDER_BY_ID.join(',')})"
+    end
```

Comment:
Another weird potential gotcha: the `activities` table is not actually included as part of this abstract query, and is instead joined (via different paths) in each inheritor.  This introduces a bit of undocumented magic that I'd be glad to eliminate if anyone has a good suggestion.  (We join via different paths for different queries because sometimes this join comes from `activity_sessions`, but sometimes it comes from a join on `follow_up_activity_id` on separate `activities` join when trying to find post diagnostic information or through `recommendations` when finding follow up activities.)